File: packages/react-native-storybook/ios/SherloModule.m
#import "SherloModule.h"
#import "SherloModuleCore.h"
#import <React/RCTUtils.h>
#import <React/RCTUIManagerUtils.h>
#import <React/RCTBridge.h>

@implementation SherloModule

RCT_EXPORT_MODULE(SherloModule)

@synthesize bridge = _bridge;

static SherloModuleCore *core;

/**
 * Required method to initialize the module.
 * Creates the core instance and initializes it.
 */
- (instancetype)init {
    self = [super init];
    if (self) {
        core = [[SherloModuleCore alloc] init];
    }
    return self;
}

/**
 * Indicates that this module should be initialized on the main thread.
 */
+ (BOOL)requiresMainQueueSetup {
    return YES;
}

// Specifies the dispatch queue on which the module's methods should be executed.
- (dispatch_queue_t)methodQueue {
    return RCTGetUIManagerQueue();
}

#pragma mark - Old Architecture Methods

/**
 * Returns the Sherlo constants.
 */
- (NSDictionary *)constantsToExport
{
    return [core getSherloConstants];
}

/**
 * Toggles between Storybook and default modes.
 */
RCT_EXPORT_METHOD(toggleStorybook) {
    [core toggleStorybook:self.bridge];
}

/**
 * Explicitly switches to Storybook mode.
 */
RCT_EXPORT_METHOD(openStorybook) {
    [core openStorybook:self.bridge];
}

/**
 * Explicitly switches to default mode.
 */
RCT_EXPORT_METHOD(closeStorybook) {
    [core closeStorybook:self.bridge];
}

/**
 * Appends base64 encoded content to a file.
 */
RCT_EXPORT_METHOD(appendFile:(NSString *)path
                  withContent:(NSString *)content
                     resolver:(RCTPromiseResolveBlock)resolve
                     rejecter:(RCTPromiseRejectBlock)reject) {
    [core appendFile:path withContent:content resolver:resolve rejecter:reject];
}

/**
 * Reads a file and returns its contents as base64 encoded string.
 */
RCT_EXPORT_METHOD(readFile:(NSString *)path
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    [core readFile:path resolver:resolve rejecter:reject];
}

/**
 * Gets UI inspector data from the current view hierarchy.
 */
RCT_EXPORT_METHOD(getInspectorData:(RCTPromiseResolveBlock)resolve
                          rejecter:(RCTPromiseRejectBlock)reject) {
    [core getInspectorData:resolve rejecter:reject];
}

/**
 * Checks UI stability by comparing screenshots taken over a specified interval.
 * The saveScreenshots parameter is accepted for cross-platform compatibility but not used in iOS.
 */
RCT_EXPORT_METHOD(stabilize:(NSInteger)requiredMatches
                 intervalMs:(NSInteger)intervalMs
                  timeoutMs:(NSInteger)timeoutMs
                  saveScreenshots:(BOOL)saveScreenshots
                   resolver:(RCTPromiseResolveBlock)resolve
                   rejecter:(RCTPromiseRejectBlock)reject) {
    [core stabilize:requiredMatches intervalMs:intervalMs timeoutMs:timeoutMs resolver:resolve rejecter:reject];
}

#pragma mark - New Architecture

#ifdef RCT_NEW_ARCH_ENABLED
#if __has_include(<SherloModuleSpec/SherloModuleSpec.h>)

// New Architecture implementation for getSherloConstants
- (NSDictionary *)getSherloConstants {
    return [core getSherloConstants];
}

// New Architecture implementation for hello
- (NSString *)hello:(NSString *)name
{
  return [NSString stringWithFormat:@"Hello, %@!", name];
}

// New Architecture implementation for toggleStorybook
- (void)toggleStorybook
{
  [core toggleStorybook:self.bridge];
}

// New Architecture implementation for openStorybook
- (void)openStorybook
{
  [core openStorybook:self.bridge];
}

// New Architecture implementation for closeStorybook
- (void)closeStorybook
{
  [core closeStorybook:self.bridge];
}

// New Architecture implementation for appendFile
- (void)appendFile:(NSString *)path
      withContent:(NSString *)content
         resolver:(RCTPromiseResolveBlock)resolve
         rejecter:(RCTPromiseRejectBlock)reject
{
  [core appendFile:path withContent:content resolver:resolve rejecter:reject];
}

// New Architecture implementation for readFile
- (void)readFile:(NSString *)path
        resolver:(RCTPromiseResolveBlock)resolve
        rejecter:(RCTPromiseRejectBlock)reject
{
  [core readFile:path resolver:resolve rejecter:reject];
}

// New Architecture implementation for getInspectorData
- (void)getInspectorData:(RCTPromiseResolveBlock)resolve
                rejecter:(RCTPromiseRejectBlock)reject
{
  [core getInspectorData:resolve rejecter:reject];
}

// New Architecture implementation for stabilize
// The saveScreenshots parameter is accepted for cross-platform compatibility but not used in iOS
- (void)stabilize:(NSInteger)requiredMatches
       intervalMs:(NSInteger)intervalMs
        timeoutMs:(NSInteger)timeoutMs
  saveScreenshots:(BOOL)saveScreenshots
         resolver:(RCTPromiseResolveBlock)resolve
         rejecter:(RCTPromiseRejectBlock)reject
{
  [core stabilize:requiredMatches intervalMs:intervalMs timeoutMs:timeoutMs resolver:resolve rejecter:reject];
}

// Required implementation for Turbo Native Module
- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params
{
  return std::make_shared<facebook::react::NativeSherloModuleSpecJSI>(params);
}
#endif
#endif

@end 

File: packages/react-native-storybook/android/src/oldarch/java/io/sherlo/storybookreactnative/SherloModule.java
package io.sherlo.storybookreactnative;

// Android Framework Imports
import android.app.Activity;

// React Native Bridge Imports
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableMap;

// Java Utility and IO Imports
import java.util.Map;

/**
 * Legacy implementation of SherloModule for React Native for old architecture.
 * Native module that exposes Sherlo functionality to JavaScript.
 * Acts as a thin wrapper around SherloModuleCore which contains the actual implementation.
 */
public class SherloModule extends ReactContextBaseJavaModule {
    public static final String NAME = "SherloModule";
    private final SherloModuleCore moduleCore;

    /**
     * Initializes the module with the React context and creates the core implementation.
     *
     * @param reactContext The React Native application context
     */
    public SherloModule(ReactApplicationContext reactContext) {
        super(reactContext);
        Activity activity = getCurrentActivity();
        this.moduleCore = new SherloModuleCore(reactContext, activity);
    }

    /**
     * Returns the name of this module for React Native.
     *
     * @return The module name ("SherloModule")
     */
    @Override
    public String getName() {
        return NAME;
    }

    /**
     * Exposes constants to JavaScript, including mode, config, and state information.
     *
     * @return A map of constants for the JavaScript side
     */
    @Override
    public Map<String, Object> getConstants() {
        return moduleCore.getSherloConstants().toHashMap();
    }

    // ==== Storybook Methods ====

    /**
     * Toggles between Storybook and default mode.
     */
    @ReactMethod
    public void toggleStorybook() {
        moduleCore.toggleStorybook();
    }

    /**
     * Explicitly switches to Storybook mode.
     */
    @ReactMethod
    public void openStorybook() {
        moduleCore.openStorybook();
    }

    /**
     * Explicitly switches to default mode.
     */
    @ReactMethod
    public void closeStorybook() {
        moduleCore.closeStorybook();
    }

    // ==== File System Methods ====

    /**
     * Appends base64 encoded content to a file.
     *
     * @param filename The name of the file to append to
     * @param base64Content The base64 encoded content to append
     * @param promise Promise to resolve when the operation is complete
     */
    @ReactMethod
    public void appendFile(String filename, String base64Content, Promise promise) {
        moduleCore.appendFile(filename, base64Content, promise);
    }

    /**
     * Reads a file and returns its content as a base64 encoded string.
     *
     * @param filename The name of the file to read
     * @param promise Promise to resolve with the file content
     */
    @ReactMethod
    public void readFile(String filename, Promise promise) {
        moduleCore.readFile(filename, promise);
    }

    // ==== Inspector Methods ====

    /**
     * Gets UI inspector data from the current view hierarchy.
     *
     * @param promise Promise to resolve with the inspector data
     */
    @ReactMethod
    public void getInspectorData(Promise promise) {
        Activity activity = getCurrentActivity();
        moduleCore.getInspectorData(activity, promise);
    }

    /**
     * Checks if the UI is stable by comparing consecutive screenshots.
     *
     * @param requiredMatches The number of consecutive matching screenshots needed
     * @param minScreenshotsCount The minimum number of screenshots to take when checking for stability
     * @param intervalMs The interval between each screenshot in milliseconds
     * @param timeoutMs The overall timeout in milliseconds
     * @param saveScreenshots Whether to save screenshots to the file system
     * @param threshold Matching threshold (0.0 to 1.0); smaller values are more sensitive
     * @param includeAA If false, ignore anti-aliased pixels when counting differences
     * @param promise Promise to resolve with true if UI becomes stable, false if timeout occurs
     */
    @ReactMethod
    public void stabilize(int requiredMatches, int minScreenshotsCount, int intervalMs, int timeoutMs, boolean saveScreenshots, double threshold, boolean includeAA, Promise promise) {
        Activity activity = getCurrentActivity();
        moduleCore.stabilize(activity, requiredMatches, minScreenshotsCount, intervalMs, timeoutMs, saveScreenshots, threshold, includeAA, promise);
    }

    /**
     * Detects if the currently visible screen can be vertically scrolled for long-screenshot capture.
     *
     * @param promise Promise to resolve with boolean (true if scrollable, false otherwise)
     */
    @ReactMethod
    public void isScrollableSnapshot(Promise promise) {
        Activity activity = getCurrentActivity();
        moduleCore.isScrollableSnapshot(activity, promise);
    }

    /**
     * Deterministically scrolls to a checkpoint index.
     *
     * @param index The checkpoint index (0-based)
     * @param offset The vertical offset per checkpoint (in pixels)
     * @param maxIndex The maximum allowed index
     */
    @ReactMethod
    public void scrollToCheckpoint(double index, double offset, double maxIndex, Promise promise) {
        Activity activity = getCurrentActivity();
        moduleCore.scrollToCheckpoint(activity, index, offset, maxIndex, promise);
    }
}
